common/captcha/CaptchaSolverInterface.py

```diff
@@ -0,0 +1,90 @@
+from abc import ABC, abstractmethod
+import os
+import cv2
+import time
+
+
+class SolvedCaptcha:
+    """
+    A class interface for representing solved Captchas.
+    """
+
+    def __init__(self, image, text, answer=None, out_dir=None):
+        """
+        :param image: Unprocessed Captcha image
+        :param text: The text contained in the captcha image. (eg: 12+3=)
+        :param answer: The solved captcha's answer (eg: 15). In the 'jumbled text' captchas this is the same as 'text'.
+        :param out_dir: Where to write captchas to upon notify_correct() or notify_incorrect()
+        """
+        self.image = image
+        self.text = text
+        self.answer = answer
+        self.out_dir = out_dir
+
+    def notify_correct(self):
+        """
+        Writes the Captcha's image to self.correct_dir
+        This should be called in the case a captcha is solved correctly.
+        Captchas should be solved with a unique filename indicating what the captcha says.
+        Eg: YYYYMMDD-HHMMSS_qwerty.png
+        """
+        try:
+            correct_dir = os.path.join(self.out_dir, 'correct')
+            os.makedirs(correct_dir, exist_ok=True)
+            filename = '{}_{}.png'.format(time.strftime("%Y%m%d-%H%M%S"), self.text)
```

Comment:
I can see cases where we'd succeed/fail twice within the same second if we have multiple scrapers running with the same outdir. Can we include more precision? A simple approach like the following would work well:
`"%s_%s.png" % (datetime.datetime.isoformat(datetime.datetime.now()), self.text)`

IMO `2020-06-12T21:18:33.643932.png` is also a little easier to visually parse than `20200612-211833.png`.

Reply:
Yes I agree that would probably avoid future headaches.

Reply:
As mentioned before the colons aren't allowed. To get around the current limited format I'll do this:

`datetime.now().strftime("%Y-%m-%dT%H-%M-%S.%f")[:-3]`

Which gives times that look like this: `2020-06-14T00-10-44.601`

Combined with the captcha text, this looks like: `2020-06-14T00-10-44.601_46+2=.png` for a captcha saying `46+2=`.

Reply:
LGTM.